File: packages/ts-host/src/build-task.ts
import { v4 as uuidv4 } from 'uuid'
import type { Message, Task } from '@a2a-js/sdk'
import { filterA2uiByCatalog, type OutputNegotiation } from './output-modes'
import { toA2uiSnapshot } from './a2ui'
import type { A2uiComponent, A2uiSnapshot, AgentResult } from './types'

const now = (): string => new Date().toISOString()

/**
 * Инвариант a2ui-action-owner-by-surface: каждый `input-required` элемент A2UI уезжает КОНВЕРТОМ
 * `A2uiSnapshot` с `surfaceId` — оркестратор строит по нему durable-маппинг «surface →
 * агент-владелец» и маршрутизирует сабмит формы именно её владельцу. Сырое дерево (включая
 * `followup` — путь elevator'а) нормализуется в конверт; дефолт id выводится из taskId: стабилен
 * между шагами ОДНОГО визарда (resume того же таска) и уникален между визардами/повторными
 * запусками в диалоге (новый запуск = новый таск). Заданные агентом id не трогаются (сквозной
 * контракт lookup/in-place replace).
 */
function ensureEnvelopeSurfaceIds(
  items: (A2uiComponent | A2uiSnapshot)[],
  taskId: string,
): A2uiSnapshot[] {
  let minted = 0
  return items.map((item) => {
    const envelope = toA2uiSnapshot(item)
    if (envelope.surfaceId) return envelope
    minted += 1
    return { ...envelope, surfaceId: minted === 1 ? `surf-${taskId}` : `surf-${taskId}-${minted}` }
  })
}

/** Дефолт без негоциации: текст-only (каталог не согласован → A2UI не шлём). */
const TEXT_ONLY: OutputNegotiation = { text: 'text/plain', catalogIds: [], catalogId: null }

export function agentMessage(
  taskId: string,
  contextId: string,
  text: string,
): Message {
  return {
    kind: 'message',
    messageId: uuidv4(),
    role: 'agent',
    parts: [{ kind: 'text', text }],
    contextId,
    taskId,
  }
}

/**
 * Заворачивает результат handler'а в A2A-`Task`. `negotiation` определяет content-negotiation
 * вывода (РЕШЕНИЕ 10, две оси): A2UI (включая HITL-карточку `followup`) — только если каталог
 * согласован (`negotiation.catalogId`); текст для `completed` — только если агент дал `message`
 * (никаких дефолтов). По умолчанию (без negotiation) — text-only.
 */
export function toTask(
  result: AgentResult,
  taskId: string,
  contextId: string,
  negotiation: OutputNegotiation = TEXT_ONLY,
): Task {
  // A2UI отдаётся только для согласованных каталогов (per-component роутинг); иначе пусто (агент даёт текст).
  // Компоненты остаются СЫРЫМИ деревьями (`{component, props, children?, catalogId?}`) — уплощение в
  // операции делает потребитель через `componentToA2uiOperations` (так оркестратор может пробросить их выше).
  // Конверты `A2uiSnapshot` проходят ЦЕЛИКОМ (сквозной контракт lookup: relay-оркестратор кладёт их
  // в свой result.a2ui, его host эмитит с теми же id); фильтр каталога — по вложенному компоненту.
  const a2ui = (result.a2ui ?? []).filter(
    (item) => filterA2uiByCatalog([toA2uiSnapshot(item).component], negotiation).length > 0,
  )
  const followup =
    result.followup && negotiation.catalogIds.includes(result.followup.catalogId ?? negotiation.catalogId ?? '')
      ? result.followup
      : undefined

  if (result.status === 'failed') {
    return {
      kind: 'task',
      id: taskId,
      contextId,
      status: {
        state: 'failed',
        message: agentMessage(taskId, contextId, result.message ?? 'Ошибка'),
        timestamp: now(),
      },
    }
  }

  if (result.status === 'input-required') {
    return {
      kind: 'task',
      id: taskId,
      contextId,
      status: {
        state: 'input-required',
        message: agentMessage(taskId, contextId, result.message ?? 'Уточните'),
        timestamp: now(),
      },
      metadata: {
        // Формы уезжают конвертами с гарантированным surfaceId (см. ensureEnvelopeSurfaceIds).
        a2ui: ensureEnvelopeSurfaceIds(followup ? [followup] : a2ui, taskId),
        ...(result.state !== undefined ? { state: result.state } : {}),
      },
    }
  }

  return {
    kind: 'task',
    id: taskId,
    contextId,
    status: {
      state: 'completed',
      // Текст — только если агент его дал (компонент-онли каноничен: AG-UI content опционален,
      // A2A не требует текстовый part). Никаких болванок '.Готово'.
      ...(result.message
        ? { message: agentMessage(taskId, contextId, result.message) }
        : {}),
      timestamp: now(),
    },
    ...(result.state !== undefined ? { metadata: { state: result.state } } : {}),
    artifacts: [
      {
        artifactId: uuidv4(),
        name: 'result',
        parts: [
          {
            kind: 'data',
            data: { a2ui, result: result.result },
          },
        ],
      },
    ],
  }
}

File: packages/python-host/src/ai37_agent_host/build_task.py
"""Сборка A2A-артефактов из ``AgentResult`` — порт ``ts-host/src/build-task.ts``.

В отличие от TS (где строится Task-dict и публикуется в bus), ``a2a-sdk`` 1.x финализирует
таск через ``TaskUpdater`` (см. ``a2a_executor``). Здесь — только чистые хелперы: сборка
protobuf-``Part`` и content-negotiation результата (РЕШЕНИЕ 10, две оси).
"""

from __future__ import annotations

from typing import Any

from a2a.types import Part
from google.protobuf.json_format import ParseDict

from .output_modes import filter_a2ui_by_catalog
from .types import A2uiComponent, AgentResult, OutputNegotiation


def text_part(text: str) -> Part:
    """protobuf text-``Part`` (``ParseDict`` принимает camelCase → ``media_type``)."""
    return ParseDict({"text": text, "mediaType": "text/plain"}, Part())


def data_part(data: dict[str, Any]) -> Part:
    """protobuf data-``Part`` (JSON Struct)."""
    return ParseDict({"data": data, "mediaType": "application/json"}, Part())


def component_to_dict(component: A2uiComponent) -> dict[str, Any]:
    """Сырое дерево-компонент → JSON-dict (``{component, props, id?, catalogId?, children?}``).

    Уплощение в v0.9-операции делает потребитель (``component_to_a2ui_operations``) — здесь
    сохраняем СЫРОЕ дерево, как TS ``toTask`` (так оркестратор может пробросить его выше).
    """
    out: dict[str, Any] = {"component": component.component, "props": component.props}
    if component.id is not None:
        out["id"] = component.id
    if component.catalog_id is not None:
        out["catalogId"] = component.catalog_id
    if component.children:
        out["children"] = {
            slot: (
                [component_to_dict(x) for x in child]
                if isinstance(child, list)
                else component_to_dict(child)
            )
            for slot, child in component.children.items()
        }
    return out


def resolve_result_a2ui(
    result: AgentResult,
    negotiation: OutputNegotiation,
) -> tuple[list[dict[str, Any]], list[dict[str, Any]]]:
    """(a2ui, followup) как СЫРЫЕ дерево-dict'ы после content-negotiation.

    A2UI (включая HITL ``followup``) — только если каталог согласован; иначе пусто.
    """
    a2ui = [component_to_dict(c) for c in filter_a2ui_by_catalog(result.a2ui, negotiation)]
    followup: list[dict[str, Any]] = []
    if result.followup is not None and negotiation.catalog_ids:
        catalog = result.followup.catalog_id or negotiation.catalog_id or ""
        if catalog in negotiation.catalog_ids:
            followup = [component_to_dict(result.followup)]
    return a2ui, followup
